File: PieGallery/app/src/main/java/com/deadely/piegallery/view/photos/view/PhotosFragment.java
package com.deadely.piegallery.view.photos.view;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.deadely.piegallery.R;
import com.deadely.piegallery.base.BaseFragment;
import com.deadely.piegallery.dataclasses.Photo;
import com.deadely.piegallery.di.component.FragmentComponent;
import com.deadely.piegallery.view.detailphoto.view.DetailPhotoActivity;
import com.deadely.piegallery.view.photos.IPhotosContract;
import com.deadely.piegallery.view.photos.PhotoAdapter;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

public class PhotosFragment extends BaseFragment implements IPhotosContract.View, PhotoAdapter.OnClickListener {
    public static String ID = "photo_id";

    private List<Photo> mPhotoResponse;
    private PhotoAdapter photoAdapter;
    @Inject
    public IPhotosContract.Presenter photosPresenter;
    private Unbinder unbinder;

    @BindView(R.id.progress_bar)
    ProgressBar progressBar;
    @BindView(R.id.rv_photo_list)
    RecyclerView rvPhotoList;
    @BindView(R.id.swipe_refresh)
    SwipeRefreshLayout swipeRefresh;
    @BindView(R.id.tv_status)
    TextView tvStatus;

    @Override
    public android.view.View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        android.view.View view = inflater.inflate(R.layout.fragment_photos, container, false);
        unbinder = ButterKnife.bind(this, view);
        initView();
        return view;
    }

    private void initView() {
        photosPresenter.attachView(this);

        GridLayoutManager LayoutManager = new GridLayoutManager(getContext(), 2);
        rvPhotoList.setLayoutManager(LayoutManager);
        photoAdapter = new PhotoAdapter(getContext());
        photoAdapter.setOnClickListener(this);
        getPhotoList();
        rvPhotoList.setAdapter(photoAdapter);

        swipeRefresh.setOnRefreshListener(() -> new Handler().postDelayed(() -> {
            swipeRefresh.setRefreshing(false);
            getPhotoList();
        }, 1000));
    }

    @Override
    public void showProgress() {
        rvPhotoList.setVisibility(GONE);
        tvStatus.setVisibility(GONE);
        progressBar.setVisibility(VISIBLE);
    }

    @Override
    public void hideProgress() {
        rvPhotoList.setVisibility(VISIBLE);
        tvStatus.setVisibility(GONE);
        progressBar.setVisibility(GONE);
    }

    private void getPhotoList() {
        photosPresenter.showProgressBar();
        photosPresenter.getPhotoList();
    }

    @Override
    public void setPhotoList(List<Photo> list) {
        mPhotoResponse = list;
        photoAdapter.setData(mPhotoResponse);

        photosPresenter.hideProgressBar();
    }

    @Override
    public void showErrorConn(String errorMessage) {
        rvPhotoList.setVisibility(GONE);
        progressBar.setVisibility(GONE);
        tvStatus.setVisibility(VISIBLE);
        tvStatus.setText(errorMessage);
    }

    @Override
    public void onClickItem(int position) {
        Intent intent = new Intent(getActivity(), DetailPhotoActivity.class);
        intent.putExtra(ID, mPhotoResponse.get(position).getId());
        startActivity(intent);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
        photosPresenter.detachView();
    }

    @Override
    protected void inject(FragmentComponent fragmentComponent) {
        fragmentComponent.inject(this);
    }
}

File: PieGallery/app/src/main/java/com/deadely/piegallery/ui/photos/view/PhotosFragment.kt
package com.deadely.piegallery.ui.photos.view

import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.appcompat.widget.SearchView
import androidx.core.content.FileProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.deadely.piegallery.R
import com.deadely.piegallery.base.BaseActivity
import com.deadely.piegallery.base.BaseFragment
import com.deadely.piegallery.dataclasses.Photo
import com.deadely.piegallery.ui.photos.PhotosAdapter
import com.deadely.piegallery.ui.photos.PhotosView
import com.deadely.piegallery.ui.photos.presenter.PhotosPresenter
import com.deadely.piegallery.utils.makeGone
import com.deadely.piegallery.utils.makeVisible
import com.paginate.Paginate
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.android.synthetic.main.fragment_photos.*
import moxy.ktx.moxyPresenter
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import javax.inject.Inject
import javax.inject.Provider

@AndroidEntryPoint
class PhotosFragment : BaseFragment(R.layout.fragment_photos), PhotosView, BaseActivity.BackButtonPressed {
    @Inject
    lateinit var presenterProvider: Provider<PhotosPresenter>

    private val presenter by moxyPresenter { presenterProvider.get() }
    private var loading = false
    private var paginator: Paginate? = null
    var search = false
    private val photosAdapter = PhotosAdapter()

    override fun onPause() {
        super.onPause()
        loading = false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.search_menu, menu)
        val menuItem = menu.findItem(R.id.mnu_search)
        menuItem.apply { setShowAsAction(MenuItem.SHOW_AS_ACTION_COLLAPSE_ACTION_VIEW or MenuItem.SHOW_AS_ACTION_IF_ROOM) }
        val searchView = menuItem.actionView as SearchView
        menuItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem?): Boolean {
                search = true
                return true
            }

            override fun onMenuItemActionCollapse(item: MenuItem?): Boolean {
                search = false
                clearList()
                presenter.page = 1
                setupPaginate()
                presenter.getPhotos()
                return true
            }
        })
        with(searchView) {
            setOnQueryTextListener(
                object : SearchView.OnQueryTextListener {
                    override fun onQueryTextSubmit(p0: String?): Boolean {
                        presenter.searchQuery(p0)
                        rvPhotoList.smoothScrollToPosition(0)
                        searchView.clearFocus()
                        return false
                    }

                    override fun onQueryTextChange(p0: String?): Boolean {
                        return false
                    }
                }
            )
        }
    }

    override fun initView() {
        presenter.isRefreshing = true
        photosAdapter.setListener(object : PhotosAdapter.OnClickListener {

            override fun addToFavorite(photo: Photo) {
                presenter.addToFavorite(photo)
            }

            override fun deleteFromFavorite(photo: Photo) {
                presenter.deleteFromFavorite(photo)
            }

            override fun onUserDataClick(photo: Photo) {
                presenter.openUserProfile(photo)
            }

            override fun onSharePhoto(photo: Photo, bitmap: Bitmap) {
                sharePicture(photo, bitmap)
            }
        })
        rvPhotoList.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = photosAdapter
        }
        setupPaginate()
        presenter.getPhotos()
    }

    private fun setupPaginate() {
        paginator = Paginate.with(
            rvPhotoList,
            object : Paginate.Callbacks {
                override fun onLoadMore() {
                    loading = true
                    if (!search) {
                        presenter.getPhotos()
                    } else {
                        paginator?.unbind()
                    }
                }

                override fun isLoading(): Boolean {
                    return loading
                }

                override fun hasLoadedAllItems(): Boolean {
                    if (presenter.isLoadedAllItems()) {
                        paginator?.unbind()
                        return true
                    }
                    return false
                }
            }
        ).build()
    }

    override fun setListeners() {
        swipeRefresh.setOnRefreshListener {
            swipeRefresh.isRefreshing = false
            if (!loading) {
                presenter.onRefresh()
            }
        }
    }

    private fun sharePicture(photo: Photo, bitmap: Bitmap) {
        var bmpUri: Uri? = null
        bmpUri = getBitmapFromDrawable(bitmap)
        bmpUri?.let {
            val shareIntent = Intent(Intent.ACTION_SEND).apply {
                putExtra(Intent.EXTRA_STREAM, bmpUri)
                putExtra(
                    Intent.EXTRA_TEXT,
                    "${context?.getString(R.string.app_name)}" +
                        "\n ${photo.user?.username} (instagram: ${photo.user?.instagramUsername ?: "the reference is missing"})" +
                        "\n ${photo.description ?: ""}"
                )
                type = "image/*"
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
            startActivity(Intent.createChooser(shareIntent, "Share Image"))
        } ?: run {
            // showShareError()
            return@sharePicture
        }
    }

    override fun setPhotos(list: List<Photo>) {
        loading = false
        photosAdapter.setData(list)
    }

    override fun clearList() {
        photosAdapter.clear()
    }

    override fun showProgress() {
        pbPhotos.makeVisible()
        swipeRefresh.makeGone()
        tvStatus.makeGone()
    }

    override fun hideProgress() {
        pbPhotos.makeGone()
        tvStatus.makeGone()
        swipeRefresh.makeVisible()
    }

    override fun showError() {
        tvStatus.makeVisible()
        pbPhotos.makeGone()
        swipeRefresh.makeGone()
    }

    override fun onBackButtonPressed() {
        presenter.exit()
    }

    private fun getBitmapFromDrawable(bmp: Bitmap): Uri? {
        var bmpUri: Uri? = null
        try {
            val file = File(
                activity?.getExternalFilesDir(Environment.DIRECTORY_PICTURES),
                "share_image_" + System.currentTimeMillis() + ".png"
            )
            val out = FileOutputStream(file)
            bmp.compress(Bitmap.CompressFormat.PNG, 90, out)
            out.close()
            bmpUri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                activity?.applicationContext?.let {
                    FileProvider.getUriForFile(
                        it,
                        "com.deadely.piegallery",
                        file
                    )
                }
            } else {
                Uri.fromFile(file)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return bmpUri
    }

    override fun getExtras() {
    }
}
